### ISSUE TITLE
Cannot upgrade Micronaut from Gradle 4.10 to Gradle 5 or above due to JAR locking

### ISSUE THREAD

**Issue by graemerocher** [gradle/gradle#11619]
We have been trying for some time now to upgrade to Gradle 5 or 6 and are blocked by an issue where Gradle locks the JAR files for submodules that are also annotation processors meaning if these subprojects are used as an annotation processor other subprojects they cannot be loaded due to Gradle locking them

This results in a core dump unless `-Dsun.zip.disableMemoryMapping=true` is specified however specifying this results in services not being able to be loaded that the annotation processors require.

### Expected Behavior

Gradle should not lock the JAR files of annotation processor projects in subprojects as in Gradle 4.x

### Current Behavior

Gradle locks annotation processor subprojects preventing their correct usage in other modules of the same build.

### Context

We cannot upgrade the Micronaut core build above Gradle 4.10 and therefore cannot test on JDK 13 or above.

Related PR https://github.com/micronaut-projects/micronaut-core/pull/2128

### Steps to Reproduce 

```
git clone https://github.com/zendern/micronaut-core.git
cd micronaut-core
git checkout -b zendern-gradle-upgrade master
git pull https://github.com/zendern/micronaut-core.git gradle-upgrade
./gradlew test
```

**Comment by graemerocher** [comment]
I tried to work around this in this branch by catching the exceptions in loading the services the Micronaut compiler users that are loaded via `ServiceLoader` and fallbacking back to instantiating them directly:

https://github.com/micronaut-projects/micronaut-core/commit/328f8bb02c29cca4dadbcd90a57b80b6c25f0b4b#diff-8722e88034b196b53ceb5f774c64603eR239

This allows the build to get further but something is still very broken with tests failing with SSL errors. Also if I run 

`./gradlew test --parallel --continue`

The JVM completely crashes unless I add `export GRADLE_OPTS="-Dsun.zip.disableMemoryMapping=true`

Once I added that and I run `./gradlew test --parallel --continue` then compilation errors occur:

```
Java HotSpot(TM) 64-Bit Server VM warning: ignoring option MaxPermSize=372m; support was removed in 8.0

> Task :core:compileJava
Note: Some input files use unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.

> Task :inject:compileJava
error: error reading /Users/graemerocher/dev/micronaut/core/core/build/libs/core-1.3.0.BUILD-SNAPSHOT.jar; error in opening zip file
error: error reading /Users/graemerocher/dev/micronaut/core/core/build/libs/core-1.3.0.BUILD-SNAPSHOT.jar; cannot read zip file
/Users/graemerocher/dev/micronaut/core/inject/src/main/java/io/micronaut/context/AbstractParametrizedBeanDefinition.java:21: error: package io.micronaut.core.annotation does not exist
import io.micronaut.core.annotation.AnnotationMetadata;
                                   ^
/Users/graemerocher/dev/micronaut/core/inject/src/main/java/io/micronaut/context/AbstractParametrizedBeanDefinition.java:22: error: package io.micronaut.core.annotation does not exist
import io.micronaut.core.annotation.Internal;
```

Note the output that says `error reading /Users/graemerocher/dev/micronaut/core/core/build/libs/core-1.3.0.BUILD-SNAPSHOT.jar; error in opening zip file`

So it is even locking JAR files produced via compilation somehow.

**Comment by big-guy** [comment]
@graemerocher it looks like there's been some changes since you first opened this.  What's a good commit that reproduces the locking error?  All I can get are compilation errors.

Which version of Java are you using?

Can you describe what you're seeing WRT which jars are being locked and which jars are being modified?  It seems like you're saying the _annotation processor_ jars are being modified by another project?

**Comment by graemerocher** [comment]
@big-guy Thanks for the response. Can you try:

```
git clone https://github.com/micronaut-projects/micronaut-core.git
cd micronaut-core
git checkout zendern-gradle-upgrade
./gradlew clean cJ --no-daemon
```

You should be output like the following for the `runtime` subproject:

```
Failed to configure default AnnotationMappers. Using fallback behaviour: Failed to load resources for service [io.micronaut.inject.annotation.AnnotationMapper]: jar:file:/Users/graemerocher/dev/micronaut/core/inject/build/libs/inject-1.3.0.BUILD-SNAPSHOT.jar!/META-INF/services/io.micronaut.inject.annotation.AnnotationMapper
Failed to configure default AnnotationMappers. Using fallback behaviour: Failed to load resources for service [io.micronaut.inject.annotation.AnnotationRemapper]: jar:file:/Users/graemerocher/dev/micronaut/core/inject/build/libs/inject-1.3.0.BUILD-SNAPSHOT.jar!/META-INF/services/io.micronaut.inject.annotation.AnnotationRemapper
Failed to configure default TypeElementVisitors. Using fallback behaviour: Failed to load resources for service [io.micronaut.inject.visitor.TypeElementVisitor]: jar:file:/Users/graemerocher/dev/micronaut/core/inject/build/libs/inject-1.3.0.BUILD-SNAPSHOT.jar!/META-INF/services/io.micronaut.inject.visitor.TypeElementVisitor
```

Which is one manifestation of the problem. If you run:

```
./gradlew clean cJ --parallel --no-daemon
```

You get compilation errors which are unexpected with the `inject` project unable to read the classes of the `core` project even though the jar contains theses classes:

```
> Task :inject:compileJava
error: error reading /Users/graemerocher/dev/micronaut/core/core/build/libs/core-1.3.0.BUILD-SNAPSHOT.jar; error in opening zip file
error: error reading /Users/graemerocher/dev/micronaut/core/core/build/libs/core-1.3.0.BUILD-SNAPSHOT.jar; cannot read zip file
/Users/graemerocher/dev/micronaut/core/inject/src/main/java/io/micronaut/context/AbstractParametrizedBeanDefinition.java:21: error: package io.micronaut.core.annotation does not exist
import io.micronaut.core.annotation.AnnotationMetadata;
                                   ^
```

> Which version of Java are you using?

We build with JDK 8.

> 
> Can you describe what you're seeing WRT which jars are being locked and which jars are being modified? It seems like you're saying the _annotation processor_ jars are being modified by another project?

Let me try and describe the issue better. The `micronaut-core` project is a multi project build. One of the subprojects is an annotation processor. This is `micronaut-inject-java`: https://github.com/micronaut-projects/micronaut-core/tree/master/inject-java

The `inject-java` subproject is placed on the annotation processor classpath of other subprojects. See for example the `runtime` subproject: https://github.com/micronaut-projects/micronaut-core/blob/zendern-gradle-upgrade/runtime/build.gradle#L42

With Gradle 4.10 this was not a problem. But with Gradle 5 and above what happens is that the annotation processor is unable to load, via service loader, files and classes from JARs on the annotation processor classpath since this JARs are locked somehow.

This manifests in a few places that breaks our build. One place is in `AbstractAnnotationMetadataBuilder` which uses a variation of service loader to load classes of type `AnnotationMapper`: https://github.com/micronaut-projects/micronaut-core/blob/zendern-gradle-upgrade/inject/src/main/java/io/micronaut/inject/annotation/AbstractAnnotationMetadataBuilder.java#L58

In Gradle 4.10 this worked, but in Gradle 5+ this results in a `ServiceConfigurationError` and I placed a hack to fallback to statically loading this classes instead of dynamically which is here: https://github.com/micronaut-projects/micronaut-core/blob/zendern-gradle-upgrade/inject/src/main/java/io/micronaut/inject/annotation/AbstractAnnotationMetadataBuilder.java#L79

You can reproduce this with the following steps. If you uncomment the following lines which allows attaching a remote debugger to Gradle compilation: https://github.com/micronaut-projects/micronaut-core/blob/zendern-gradle-upgrade/runtime/build.gradle#L56

Then set a break point on https://github.com/micronaut-projects/micronaut-core/blob/zendern-gradle-upgrade/inject/src/main/java/io/micronaut/inject/annotation/AbstractAnnotationMetadataBuilder.java#L58

And run the following command:

```
./gradlew clean cJ --no-daemon
```

And attach the debugger you will see that when the `runtime` project is compiled the annotation processor is able to find the `META-INF/services/io.micronaut.inject.annotation.AnnotationMapper` files within the JAR files that are on the classpath of the annotation processor, however when it attempts to load any of these files via `URL.openStream()` the result is a `FileNotFoundException` which then manifests as the `ServiceConfigurationError` which then results in output like:

```
Failed to configure default AnnotationMappers. Using fallback behaviour: Failed to load resources for service [io.micronaut.inject.annotation.AnnotationMapper]: jar:file:/Users/graemerocher/dev/micronaut/core/inject/build/libs/inject-1.3.0.BUILD-SNAPSHOT.jar!/META-INF/services/io.micronaut.inject.annotation.AnnotationMapper
```

This did not occur in Gradle 4.10 and the reason why it is occuring in Gradle 5+ doesn't make sense to me. It seems that Gradle 5+ is somehow preventing the annotation processor (possibly by not closes a reference to these JAR files) from reading the contents of any JAR file produced by the same build that is on the annotation processor classpath. I hope this serves as a better explanation of the project.

Note that I have tried various things such as removing the usage of shadow JAR (too see if that was the problem) but it does not appear to be.

**Comment by big-guy** [comment]
Thanks for all the details @graemerocher.

It looks like this is being caused by the shadowJar task overwriting the existing jar concurrently to compilation by other projects.  I got a little suspicious when I was able to get successful builds by running fewer tasks or running mostly up-to-date builds.

I then noticed this:
https://scans.gradle.com/s/ues2t7w3x5qe4/timeline?details=65ev7keldv2bw

The shadowJar task was running at the same time as the failing compile task. To dig a little deeper, I disable the shadow task for core and tried again:
https://scans.gradle.com/s/q2skbc6c4anfu/timeline?details=742yun4u67p3c&end=1576046852382&start=1576046851634

In this case, inject compiled fine, but the problem moved to another project that also had another shadowJar running concurrently.

I then checked the configuration of the shadowJar and saw that it was writing to the same location as the regular jar task.  This sounded familiar, so I tried bumping to the latest Shadow plugin (5.2.0), but that didn't fix it.  So I removed this line: https://github.com/micronaut-projects/micronaut-core/blob/zendern-gradle-upgrade/build.gradle#L695 which changed the path to the jar to something else.  I think what may have changed between 4.x and now is how this default value is determined and the shadow plugin is inadvertently using the same as the regular jar task.

This got me through more of the compilation, up to this point:
https://scans.gradle.com/s/wjmassgsie726/failure?openFailures=WzBd&openStackTraces=WzNd#top=0

This looks like a mismatch between something expecting the shaded version of ASM and getting the non-shaded version.  Does this ring any bells?

**Comment by big-guy** [comment]
@lptr is this something we could detect somehow?  e.g., if a task declares that it writes to a file that's an input to an executing task, we should warn/fail because this means something is wrong.  That could be a missing dependency or two tasks overwriting their outputs (like in this case)

**Comment by graemerocher** [comment]
@big-guy That would be an indication that classes have been rewritten to use the shaded version of ASM but the `core` module which contains these shaded versions hasn't been rewritten and therefore those classes are not the classpath

**Comment by lptr** [comment]
@big-guy we don't do checks like that yet. The plan for next year is add validation to the task graph, so if a task uses a (part of the) output produced by another task, then there has to be a build dependency between them. That should cover the use case you mentioned.

**Comment by graemerocher** [comment]
@big-guy Thanks for sending me in the right direction. I have managed to progress the upgrade further. I have no idea how you were able to discover the cause, I guess one needs to be an expert on Gradle internals to figure out these things. 

however I am now quite frequently hitting https://github.com/gradle/gradle/issues/10985 which results in the need for a full recompilation. One comment on that issue says it is fixed in 6.0.1 but that does not appear to be the case for me.

**Comment by big-guy** [comment]
>  I have no idea how you were able to discover the cause, I guess one needs to be an expert on Gradle internals to figure out these things.

The symptoms looked like something was "corrupting" the jar during compilation.  I've been burned by that before (outside Gradle).  The timeline in https://scans.gradle.com/s/ues2t7w3x5qe4/timeline?details=qalcau5375bxi made me look at core:shadowJar first:
![image](https://user-images.githubusercontent.com/5387972/70682268-52d5a400-1c6c-11ea-96d1-a6d590487eb1.png)


I posted a comment with a nightly to try for your other issue.  I'll close this issue for now.
